src/odemis/driver/test/tfsbc_test.py

```diff
@@ -0,0 +1,92 @@
+# -*- coding: utf-8 -*-
+"""
+Created on 11 May 2020
+
+@author: Philip Winkler
+
+Copyright © 2020 Philip Winkler, Delmic
+
+This file is part of Odemis.
+
+Odemis is free software: you can redistribute it and/or modify it under the terms
+of the GNU General Public License version 2 as published by the Free Software
+Foundation.
+
+Odemis is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
+PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+Odemis. If not, see http://www.gnu.org/licenses/.
+"""
+
+from odemis.driver import tfsbc
+from odemis import model
+import unittest
+import logging
+import os
+import time
+import math
+
+TEST_NOHW = 1#(os.environ.get("TEST_NOHW", 0) != 0)  # Default to Hw testing
+
+logging.getLogger().setLevel(logging.DEBUG)
+logging.basicConfig(format="%(asctime)s  %(levelname)-7s %(module)s:%(lineno)d %(message)s")
+
+if TEST_NOHW:
+    PORT = "/dev/fake"
+else:
+    PORT = "/dev/ttyACM*"
+
+# @skip("skip")
+class TestBeamController(unittest.TestCase):
+    """
+    Tests the beam controller driver.
+    """
+
+    @classmethod
+    def setUpClass(cls):
+        cls.bc = tfsbc.BeamController("DC Offset", None, PORT)
+        cls.bc._metadata[model.MD_CALIB_SCALE] = [3e6, 3e6]  # 3e6 amp/m = 3 mamp / nm
+        cls.bc._metadata[model.MD_CALIB_TRANSLATION] = [0, 0]
+        cls.bc._metadata[model.MD_CALIB_ROTATION] = 0  # rad
+
+    @classmethod
+    def tearDownClass(cls):
+        pass
+
+    def assertTupleAlmostEqual(self, first, second, places=None, msg=None, delta=None):
+        """
+        check two tuples are almost equal (value by value)
+        """
+        for f, s in zip(first, second):
+            self.assertAlmostEqual(f, s, places=places, msg=msg, delta=delta)
+
+    def test_read_write(self):
+        vals = [27000, 37000, 20000, 44000]
+        self.bc._write_registers(vals)
+        ret = self.bc._read_registers()
+        self.assertTupleAlmostEqual(vals, list(ret), places=1)
+
+    def test_set_shift(self):
+        startt = time.time()
+        shift = (5e-9, 8e-9)
+        self.bc.shift.value = shift
+        self.assertLess(time.time() - startt, 0.01, "Reading/writing took more than 10 ms.")
+        logging.debug("Shift value set to %s", self.bc.shift.value)
+        self.assertTupleAlmostEqual(self.bc.shift.value, shift, places=9)
+
+    def test_rotation(self):
+        """
+        Test if setting shift works with non-zero rotation.
+        """
+        self.bc._metadata[model.MD_CALIB_ROTATION] = math.pi / 4  # rad
+        shift = (5e-9, 8e-9)
+        self.bc.shift.value = shift
+        self.assertTupleAlmostEqual(self.bc.shift.value, shift, places=9)
+        self.bc._metadata[model.MD_CALIB_ROTATION] = 0  # rad
```

Comment:
If you change back the rotation, does the shift than also update? If yes, you could also test that.

Reply:
No, it shouldn't update. It's just to put it back for the next testcase.

Reply:
Are you sure that's a good idea? I think the .shift VA should reflect the state of the hardware. I would suggest do nothing when updating the metadata and only apply the rotation at the next change of .shift.
What do you think?

Reply:
Well, at least we agree that .shift should reflect the state of the hardware. This "reflection" takes into account the calibration values. So if you update the calibration values, the .shift value should change. If you don't do this, you get some weird behaviour (which I have experienced in some other software). Like, you set the shift to a value X, => the hardware goes to position X'. Then you update the calibration (nothing else happens). You read the shift, it's still value X. Finally, you set the shift to value X *again*, and suddenly the hardware moves to position X" ! Reading an attribute and immediately setting it to the same value shouldn't cause any hardware change.

So, for the driver to behave nicely, I think that after the calibration is updated, the shift value should be computed back based on the hardware value + the calibration.

Reply:
I see your point! I'll make the change.

Actually, how do I do this? I guess I'll have to overwrite updateMetadata and from there change the VA. But I can't call the setter, otherwise the value will be written to the hardware registers. Is there a way to modify the value without triggering the setter function?

Reply:
Just to make sure I understand completely: When we write to shift.value, we enter the value in the stage coordinate system without offset. Then, when reading the value back, it's going to be in the beam coordinate system (rotation), and with the offset (and modified slightly to account for the rounding when converting to int16).
So we might have something like
```
controller.shift.value = (5e-9, 3e-9)
print(controller.shift.value) --> will output (8.32e-9, 2.4e-9)
```
(completely made up numbers)

Reply:
If you read the previous comment I just deleted, please ignore, I misunderstood.